File: src/test/scala/com/exasol/cloudetl/parquet/ParquetWriteOptionsTest.scala
package com.exasol.cloudetl.parquet

import com.exasol.cloudetl.storage.StorageProperties

import org.apache.parquet.hadoop.ParquetWriter
import org.apache.parquet.hadoop.metadata.CompressionCodecName
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite

class ParquetWriteOptionsTest extends AnyFunSuite with BeforeAndAfterEach {

  private[this] var properties: Map[String, String] = _

  override final def beforeEach(): Unit = {
    properties = Map.empty[String, String]
    ()
  }

  test("apply returns default values") {
    val options = ParquetWriteOptions(StorageProperties(properties))
    assert(options.compressionCodec === CompressionCodecName.UNCOMPRESSED)
    assert(options.blockSize === ParquetWriter.DEFAULT_BLOCK_SIZE)
    assert(options.pageSize === ParquetWriter.DEFAULT_PAGE_SIZE)
    assert(options.enableDictionaryEncoding === true)
    assert(options.enableValidation === true)
  }

  test("apply returns user provided compression codec") {
    val testData = Map(
      "snappy" -> CompressionCodecName.SNAPPY,
      "gzip" -> CompressionCodecName.GZIP,
      "lzo" -> CompressionCodecName.LZO,
      "other" -> CompressionCodecName.UNCOMPRESSED
    )
    testData.foreach { case (given, expected) =>
      properties = Map("PARQUET_COMPRESSION_CODEC" -> given)
      assert(ParquetWriteOptions(StorageProperties(properties)).compressionCodec === expected)
    }
  }

  test("apply returns user provided block size") {
    properties = Map("PARQUET_BLOCK_SIZE" -> "64")
    assert(ParquetWriteOptions(StorageProperties(properties)).blockSize === 64)
  }

  test("apply throws if block size value cannot be converted to integer") {
    properties = Map("PARQUET_BLOCK_SIZE" -> "6l4")
    val thrown = intercept[NumberFormatException] {
      ParquetWriteOptions(StorageProperties(properties))
    }
    assert(thrown.getMessage === s"""For input string: "6l4"""")
  }

  test("apply returns user provided page size") {
    properties = Map("PARQUET_PAGE_SIZE" -> "128")
    assert(ParquetWriteOptions(StorageProperties(properties)).pageSize === 128)
  }

  test("apply throws if page size value cannot be converted to integer") {
    properties = Map("PARQUET_PAGE_SIZE" -> "12e")
    val thrown = intercept[NumberFormatException] {
      ParquetWriteOptions(StorageProperties(properties))
    }
    assert(thrown.getMessage === s"""For input string: "12e"""")
  }

  test("apply returns user provided dictionary encoding enabled") {
    properties = Map("PARQUET_DICTIONARY_ENCODING" -> "false")
    assert(ParquetWriteOptions(StorageProperties(properties)).enableDictionaryEncoding === false)
  }

  test("apply throws if dictionary encoding enable cannot be converted to boolean") {
    properties = Map("PARQUET_DICTIONARY_ENCODING" -> "nay")
    val thrown = intercept[IllegalArgumentException] {
      ParquetWriteOptions(StorageProperties(properties))
    }
    assert(thrown.getMessage === s"""For input string: "nay"""")
  }

  test("apply returns user provided validation enabled") {
    properties = Map("PARQUET_VALIDATION" -> "false")
    assert(ParquetWriteOptions(StorageProperties(properties)).enableValidation === false)
  }

  test("apply throws if validation enable cannot be converted to boolean") {
    properties = Map("PARQUET_VALIDATION" -> "yay")
    val thrown = intercept[IllegalArgumentException] {
      ParquetWriteOptions(StorageProperties(properties))
    }
    assert(thrown.getMessage === s"""For input string: "yay"""")
  }

}

File: src/test/java/com/exasol/cloudetl/parquet/ParquetWriteOptionsTest.java
package com.exasol.cloudetl.parquet;

import static org.junit.jupiter.api.Assertions.*;

import java.util.Map;
import java.util.stream.Stream;

import org.apache.parquet.hadoop.ParquetWriter;
import org.apache.parquet.hadoop.metadata.CompressionCodecName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import com.exasol.cloudetl.storage.StorageProperties;

class ParquetWriteOptionsTest {
    @Test
    void fromReturnsDefaultValues() {
        final ParquetWriteOptions options = ParquetWriteOptions.from(new StorageProperties(Map.of()));
        assertEquals(CompressionCodecName.UNCOMPRESSED, options.compressionCodec);
        assertEquals(ParquetWriter.DEFAULT_BLOCK_SIZE, options.blockSize);
        assertEquals(ParquetWriter.DEFAULT_PAGE_SIZE, options.pageSize);
        assertTrue(options.enableDictionaryEncoding);
        assertTrue(options.enableValidation);
    }

    @ParameterizedTest
    @MethodSource("compressionCodecs")
    void fromReturnsUserProvidedCompressionCodec(final String given, final CompressionCodecName expected) {
        assertEquals(expected,
                ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_COMPRESSION_CODEC", given))).compressionCodec);
    }

    @Test
    void fromReturnsUserProvidedBlockSize() {
        assertEquals(64,
                ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_BLOCK_SIZE", "64"))).blockSize);
    }

    @Test
    void fromThrowsIfBlockSizeValueCannotBeConvertedToInteger() {
        final NumberFormatException thrown = assertThrows(NumberFormatException.class,
                () -> ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_BLOCK_SIZE", "6l4"))));
        assertEquals("For input string: \"6l4\"", thrown.getMessage());
    }

    @Test
    void fromReturnsUserProvidedPageSize() {
        assertEquals(128,
                ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_PAGE_SIZE", "128"))).pageSize);
    }

    @Test
    void fromThrowsIfPageSizeValueCannotBeConvertedToInteger() {
        final NumberFormatException thrown = assertThrows(NumberFormatException.class,
                () -> ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_PAGE_SIZE", "12e"))));
        assertEquals("For input string: \"12e\"", thrown.getMessage());
    }

    @Test
    void fromReturnsUserProvidedDictionaryEncodingEnabled() {
        assertFalse(ParquetWriteOptions
                .from(new StorageProperties(Map.of("PARQUET_DICTIONARY_ENCODING", "false"))).enableDictionaryEncoding);
    }

    @Test
    void fromReturnsUserProvidedValidationEnabled() {
        assertFalse(ParquetWriteOptions.from(new StorageProperties(Map.of("PARQUET_VALIDATION", "false"))).enableValidation);
    }

    private static Stream<Arguments> compressionCodecs() {
        return Stream.of(//
                Arguments.of("snappy", CompressionCodecName.SNAPPY), //
                Arguments.of("gzip", CompressionCodecName.GZIP), //
                Arguments.of("lzo", CompressionCodecName.LZO), //
                Arguments.of("other", CompressionCodecName.UNCOMPRESSED));
    }
}
